staging/src/k8s.io/apimachinery/third_party/forked/golang/reflect/deep_equal.go

```diff
@@ -270,14 +274,18 @@ func (e Equalities) deepValueEqual(v1, v2 reflect.Value, visited map[visit]bool,
 // Unexported field members cannot be compared and will cause an informative panic; you must add an Equality
 // function for these types.
 func (e Equalities) DeepEqual(a1, a2 interface{}) bool {
-	return e.deepEqual(a1, a2, true)
+	return e.deepEqual(a1, a2, true, false)
 }
 
 func (e Equalities) DeepEqualWithNilDifferentFromEmpty(a1, a2 interface{}) bool {
-	return e.deepEqual(a1, a2, false)
+	return e.deepEqual(a1, a2, false, false)
+}
+
+func (e Equalities) PublicDeepEqual(a1, a2 interface{}) bool {
+	return e.deepEqual(a1, a2, true, true)
```

Comment:
I don't know if a function name per combination of options is the right way to continue to expand here..  Will we need a `PublicDeepEqualWithNilDifferentFromEmpty` later?  What if we add another independently settable option?

Reply:
Happy to have a different API than the one currently suggested, but introducing a new function without changing the call signature of `equality.Semantic.DeepEqual` is less intrusive on the [existing consumers](https://github.com/search?q=equality.Semantic.DeepEqual+k8s.io%2Fapimachinery%2Fpkg%2Fapi%2Fequality+language%3AGo&type=code&l=Go) of which there are quite a few

Reply:
We can't change the existing functions.  But if we decide to add a new function, how about a name like `DeepCopyWithOptions`?

Reply:
Yes that sounds reasonable, I propose:

We wrap the current helper method:
```
func (e Equalities) deepEqual(a1, a2 interface{}, equateNilAndEmpty, ignoreUnexportedFields bool) bool {
```
with a new method with the name you suggest: `DeepEqualWithOptions` (I assume you meant equal instead of copy)

but I do not like having a list of bools as arguments (if future options are added later), we could use a single argument with a iota of flags?, so the following type and signature:

```
type DeepEqualOpts uint8

const (
  DeepEqualEquateNilAndEmpty DeepEqualOpts = 1 << iota
  DeepEqualIgnoreUnexportedFields
)

func (e Equalities) DeepEqualWithOptions(a1, a2 interface{}, flags DeepEqualOpts) bool { 
```

and adding a new option will not break the API for existing consumers

WDYT?

Reply:
How about making the options a struct? We then just require that any change to the struct is backward compatible.  Initially it's flags that default to false, but we're not limited to flag options in the future

Reply:
Yes, sounds good, I have updated the PR with these changes